05-Mining-Protocol.md

```diff
@@ -1,101 +1,180 @@
+
+![](./img/standard_job.png)
+
+### 5.1.2 Extended Jobs
+
+Extended Jobs allow rolling Merkle Roots, giving extensive control over the search space so that they can implement various advanced use cases such as:
+- translation between SV1 and SV2 protocols
+- difficulty aggregation
+- search space splitting
+
+Extended Jobs are distributed via the [`NewExtendedMiningJob`](#5416-newextendedminingjob-server---client) message, which can be sent via:
+- [Group Channels](#533-group-channels)
+- [Extended Channels](#532-extended-channels)
+
+![](./img/extended_job.png)
+
+#### 5.1.2.1 Extended Extranonce
+
+Downstream and Upstream are relative terms. There could theoretically be multiple nested layers of proxies between a work-providing server and a mining device.
+
+So the protocol establishes the notion of **Extended Extranonce**, which is a fundamental component of Extended Jobs.
+
+The Extended Extranonce is an array of 32 bytes, split into three different areas:
 
-Whenever a standard channel is created it is always put into some channel group identified by its `group_channel_id`.
-Group channel ID namespace is the same as channel ID namespace on a particular connection but the values chosen for group channel IDs must be distinct.
+![](./img/extended_extranonce.png)
 
-### 5.1.4 Future Jobs
+- The `extranonce_prefix` bytes are reserved for the upstream layer, where fixed bytes were already established for the Extranonce and are no longer available for rolling or search space splitting.
+- The **locally reserved** bytes is where the proxy will assign unique values for its own downstream clients, according to their hashpower.
+- The **downstream reserved** bytes is where the downstream clients will further distribute the search space (if they are also proxies), or use for rolling (if they are SV1 mining devices).
 
-An empty future block job or speculated non-empty job can be sent in advance to speedup new mining job distribution.
-The point is that the mining server MAY have precomputed such a job and is able to pre-distribute it for all active channels.
-The only missing information to start to mine on the new block is the new prevhash.
+![](./img/extended_extranonce_proxies.png)
+
+In order to calculate the Merkle Root, an Extended Job carries the following data:
+- `merkle_path`
+- `coinbase_prefix`
+- `coinbase_suffix`
+
+An [Extended Channel](#532-extended-channels) has the following properties:
+- `extranonce_prefix`: the Extended Extranonce bytes that were already allocated by the upstream server.
+- `extranonce_size`: how many bytes are available for the locally reserved and downstream reserved areas of the Extended Extranonce.
+
+And a [Standard Channel](#531-standard-channels) has the following property:
+- `extranonce_prefix` the Extended Extranonce bytes that were already allocated by the upstream server.
```

Comment:
you have less messages to send from the proxy to the pool

Reply:
multiple standard channels are aggregated into a group channel

when an `ExtendedMiningJob` message comes into the group channel, it needs to be converted it into multiple `NewMiningJob` messages

each `NewMiningJob` message needs to have a unique Merkle Root, otherwise there would be collision in the search space

the uniqueness of each Merkle Root is determined by the static `extranonce_prefix` of each standard channel